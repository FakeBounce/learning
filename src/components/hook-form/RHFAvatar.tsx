import { Trans } from '@lingui/macro';
import { Avatar, Input, Stack, Typography } from '@mui/material';
import Iconify from '@src/components/iconify/Iconify';
import { uploader } from '@utils/fileUploader';
import {
  ChangeEvent,
  Dispatch,
  memo,
  ReactNode,
  SetStateAction,
  useEffect,
  useRef,
  useState
} from 'react';
import CircledAvatar from '@src/components/lms/CircledAvatar';
import { Controller, useFormContext } from 'react-hook-form';

interface RHFAvatarProps {
  name: string;
  image: string | File;
  setImage: Dispatch<SetStateAction<string | File>>;
  label?: ReactNode;
  size?: 'small' | 'medium';
}

function RHFAvatar({
  name,
  image,
  setImage,
  label = <Trans>Téléchargez une photo</Trans>,
  size = 'small'
}: RHFAvatarProps) {
  const [result, setResult] = useState<string | null>(null);
  const imageRef = useRef(null);
  const { control } = useFormContext();

  useEffect(() => {
    if (image !== '' && typeof image === 'string') {
      setResult(image);
    } else if (image instanceof File) {
      const reader = new FileReader();
      reader.onload = () => {
        setResult(reader.result as string);
      };
      reader.readAsDataURL(image);
    }
  }, [image]);

  return (
    <Controller
      name={name}
      control={control}
      render={({ field, fieldState: { error } }) => (
        <>
          <CircledAvatar>
            <Stack alignItems="center" spacing={1}>
              {result ? (
                <Avatar
                  data-testid="avatar"
                  ref={imageRef}
                  src={result}
                  sx={{
                    width: '100%',
                    height: '100%',
                    borderRadius: (theme) => theme.shape.customBorderRadius.whole
                  }}
                />
              ) : (
                <>
                  <Iconify icon="majesticons:file-plus-line" />
                  <Typography fontSize="10px">{label}</Typography>
                </>
              )}
            </Stack>
            <Input
              {...field}
              fullWidth
              value={field.value instanceof File ? field.value : ''}
              error={!!error}
              type={'file'}
              inputProps={{ accept: '.png,.jpeg,.gif,.jpg' }}
              sx={{ display: 'none' }}
              data-testid="file-upload"
              hidden
              onChange={(e: ChangeEvent<HTMLInputElement>) => {
                if (e.target.files) {
                  field.onChange(e);
                  setImage(e.target.files[0]);
                  uploader(e, setResult);
                }
              }}
              size={size}
            />
          </CircledAvatar>
          {error && (
            <Typography variant="caption" sx={{ color: (theme) => theme.palette.error.main }}>
              {error.message}
            </Typography>
          )}
        </>
      )}
    />
  );
}
export default memo(RHFAvatar);
