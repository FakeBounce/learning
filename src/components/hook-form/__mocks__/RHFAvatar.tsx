// RHFAvatar.mock.tsx
import React, { Dispatch, SetStateAction, ReactNode } from 'react';
import { Avatar, Input, Stack, Typography } from '@mui/material';
import Iconify from '@src/components/iconify/Iconify';
import CircledAvatar from '@src/components/lms/CircledAvatar';

interface RHFAvatarProps {
  name: string;
  image: string | File;
  setImage: Dispatch<SetStateAction<string | File>>;
  label?: ReactNode;
  size?: 'small' | 'medium';
}

const RHFAvatar: React.FC<RHFAvatarProps> = ({
  image,
  setImage,
  label = <span>Téléchargez une photo</span>, // Using <span> for simplicity in tests
  size = 'small'
}) => {
  const result = typeof image === 'string' ? image : null;

  return (
    <CircledAvatar>
      <Stack alignItems="center" spacing={1}>
        {result ? (
          <Avatar
            data-testid="avatar"
            src={result}
            sx={{
              width: '100%',
              height: '100%',
              borderRadius: (theme) => theme.shape.customBorderRadius.whole
            }}
          />
        ) : (
          <>
            <Iconify icon="majesticons:file-plus-line" />
            <Typography fontSize="10px">{label}</Typography>
          </>
        )}
      </Stack>
      <Input
        fullWidth
        value=""
        type="file"
        inputProps={{ accept: '.png,.jpeg,.gif,.jpg' }}
        sx={{ display: 'none' }}
        data-testid="file-upload"
        hidden
        onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
          if (e.target.files && e.target.files[0]) {
            setImage(e.target.files[0]);
          }
        }}
        size={size}
      />
    </CircledAvatar>
  );
};

export default RHFAvatar;
