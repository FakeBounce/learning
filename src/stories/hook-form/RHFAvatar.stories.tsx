import { Meta, StoryObj } from '@storybook/react';
import { FormProvider, useForm } from 'react-hook-form';
import RHFAvatar from '@src/components/hook-form/RHFAvatar';
import { useState } from 'react';

const meta: Meta = {
  title: 'Example/RHFAvatar',
  component: RHFAvatar,
  parameters: {
    backgrounds: {
      default: 'dark'
    },
    layout: 'centered'
  }
};

export default meta;

type Story = StoryObj<typeof RHFAvatar>;

const RHFAvatarTemplate: Story = {
  render: (args: any) => {
    const [image, setImage] = useState<string | File>('');
    const methods = useForm();

    return (
      <FormProvider {...methods}>
        <form>
          <RHFAvatar image={image} setImage={setImage} {...args} />
        </form>
      </FormProvider>
    );
  }
};

export const Default = {
  ...RHFAvatarTemplate,
  args: {
    name: 'default',
    label: 'Default Avatar',
    size: 'medium'
  }
};

export const WithImage = {
  ...RHFAvatarTemplate,
  args: {
    name: 'withImage',
    label: 'With Image Avatar',
    size: 'medium',
    image: 'avatar.jpg'
  }
};

export const Required = {
  ...RHFAvatarTemplate,
  args: {
    name: 'required',
    label: 'Required Avatar',
    size: 'medium',
    required: true
  }
};
