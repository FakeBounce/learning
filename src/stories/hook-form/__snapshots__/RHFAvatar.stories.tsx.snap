// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example/RHFAvatar Default smoke-test 1`] = `
<form>
  <label class="MuiBox-root css-fjsr0d">
    <div class="MuiBox-root css-jfdah8">
      <div class="MuiStack-root css-aha7cr-MuiStack-root">
        <span>
        </span>
        <p class="MuiTypography-root MuiTypography-body1 css-2z1cr3-MuiTypography-root">
          Default Avatar
        </p>
      </div>
      <div data-testid="file-upload"
           hidden
           class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth css-hd0s91-MuiInputBase-root-MuiInput-root"
      >
        <input aria-invalid="false"
               name="default"
               type="file"
               accept=".png,.jpeg,.gif,.jpg"
               class="MuiInputBase-input MuiInput-input css-pjfl94-MuiInputBase-input-MuiInput-input"
               value
        >
      </div>
    </div>
  </label>
</form>
`;

exports[`Example/RHFAvatar Required smoke-test 1`] = `
<form>
  <label class="MuiBox-root css-fjsr0d">
    <div class="MuiBox-root css-jfdah8">
      <div class="MuiStack-root css-aha7cr-MuiStack-root">
        <svg xmlns="http://www.w3.org/2000/svg"
             xmlns:xlink="http://www.w3.org/1999/xlink"
             aria-hidden="true"
             role="img"
             class="MuiBox-root css-1t9pz9x iconify iconify--majesticons"
             width="1em"
             height="1em"
             viewbox="0 0 24 24"
        >
          <path fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                d="M19 16v3m0 0v3m0-3h3m-3 0h-3m4-10v-.172a2 2 0 0 0-.586-1.414l-3.828-3.828A2 2 0 0 0 14.172 3H14m6 6h-4a2 2 0 0 1-2-2V3m6 6v3m-6-9H6a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h6"
          >
          </path>
        </svg>
        <p class="MuiTypography-root MuiTypography-body1 css-2z1cr3-MuiTypography-root">
          Required Avatar
        </p>
      </div>
      <div data-testid="file-upload"
           hidden
           class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth css-hd0s91-MuiInputBase-root-MuiInput-root"
      >
        <input aria-invalid="false"
               name="required"
               type="file"
               accept=".png,.jpeg,.gif,.jpg"
               class="MuiInputBase-input MuiInput-input css-pjfl94-MuiInputBase-input-MuiInput-input"
               value
        >
      </div>
    </div>
  </label>
</form>
`;

exports[`Example/RHFAvatar WithImage smoke-test 1`] = `
<form>
  <label class="MuiBox-root css-fjsr0d">
    <div class="MuiBox-root css-jfdah8">
      <div class="MuiStack-root css-aha7cr-MuiStack-root">
        <div class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-19blu3t-MuiAvatar-root"
             data-testid="avatar"
        >
          <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
               focusable="false"
               aria-hidden="true"
               viewbox="0 0 24 24"
               data-testid="PersonIcon"
          >
            <path d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z">
            </path>
          </svg>
        </div>
      </div>
      <div data-testid="file-upload"
           hidden
           class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth css-hd0s91-MuiInputBase-root-MuiInput-root"
      >
        <input aria-invalid="false"
               name="withImage"
               type="file"
               accept=".png,.jpeg,.gif,.jpg"
               class="MuiInputBase-input MuiInput-input css-pjfl94-MuiInputBase-input-MuiInput-input"
               value
        >
      </div>
    </div>
  </label>
</form>
`;
